src/BuildPrediction/ProjectGraphPredictionExecutor.cs

```diff
@@ -0,0 +1,152 @@
+        /// graph predictors against the resulting projects.
+        /// </summary>
+        /// <param name="graphEntryPoints">The entry points to use for the graph creation.</param>
+        /// <param name="projectCollection">The collection with which all projects in the graph should be associated.</param>
+        /// <param name="projectPredictionCollector">The prediction collector to use.</param>
+        public void PredictInputsAndOutputs(
+            IEnumerable<ProjectGraphEntryPoint> graphEntryPoints,
+            ProjectCollection projectCollection,
+            IProjectPredictionCollector projectPredictionCollector)
+        {
+            graphEntryPoints.ThrowIfNull(nameof(graphEntryPoints));
+            projectCollection.ThrowIfNull(nameof(projectCollection));
+            projectPredictionCollector.ThrowIfNull(nameof(projectPredictionCollector));
+
+            // Use a shared evaluation context for all projects.
+            EvaluationContext evaluationContext = EvaluationContext.Create(EvaluationContext.SharingPolicy.Shared);
+
+            // Note that currently we cannot control the degreee of parallelism of ProjectGraph, but if it does expose the ability
+            // to configure it, we should use _options.MaxDegreeOfParallelism.
+            var projectGraph = new ProjectGraph(
+                graphEntryPoints,
+                projectCollection,
+                (string projectPath, Dictionary<string, string> globalProperties, ProjectCollection projCollection) =>
+                {
+                    var loadSettings = ProjectLoadSettings.DoNotEvaluateElementsWithFalseCondition;
+                    var projectOptions = new ProjectOptions
+                    {
+                        GlobalProperties = globalProperties,
+                        ToolsVersion = projCollection.DefaultToolsVersion,
+                        ProjectCollection = projCollection,
+                        LoadSettings = loadSettings,
+                        EvaluationContext = evaluationContext,
+                    };
+
+                    return ProjectInstance.FromFile(projectPath, projectOptions);
+                });
+
+            // Special-case single-threaded prediction to avoid the overhead of Parallel.ForEach in favor of a simple loop.
+            if (_options.MaxDegreeOfParallelism == 1)
+            {
+                foreach (var projectNode in projectGraph.ProjectNodes)
+                {
+                    ExecuteAllPredictors(projectNode, _projectPredictors, _projectGraphPredictors, projectPredictionCollector);
+                }
+            }
+            else
+            {
+                Parallel.ForEach(
+                    projectGraph.ProjectNodes.ToArray(),
+                    new ParallelOptions() { MaxDegreeOfParallelism = _options.MaxDegreeOfParallelism },
+                    projectNode => ExecuteAllPredictors(projectNode, _projectPredictors, _projectGraphPredictors, projectPredictionCollector));
```

Comment:
There's an interesting design bug here. In the graph case, the collector loses the context on which project/node a specific prediction came from. I may need to add a `ProjectGraphPredictionReporter` and `IProjectGraphPredictionCollector` to accommodate...

Reply:
Fixed by giving collectors the associated project instance